src/tactic/core.lean

```diff
@@ -1612,6 +1612,43 @@ add_tactic_doc
 
 attribute [higher_order map_comp_pure] map_pure
 
+/-- Attribute `needed` is used to mark declarations that are featured
+in the current file.  Then, the `#unneeded` command can be used to
+list the declaration present in the file that do not support the main
+features of the file. -/
+@[user_attribute]
+meta def needed_attr : user_attribute :=
+{ name := `needed,
+  descr := "tag essential declarations to help identify unused definitions" }
+
+/-- `unneeded n m` removes from the map of unneeded declarations those
+referenced by declaration named `n` which is considerred to be a
+needed declaration -/
+private meta def unneeded : name → name_map declaration → tactic (name_map declaration)
+| n m :=
+  do d ← get_decl n,
+     if m.contains n then do
+       let m := m.erase n,
+       let ns := d.value.list_constant.union d.type.list_constant,
+       ns.mfold m unneeded
+     else pure m
+
+/-- In the current file, list all the declaration that are not marked as `@[needed]` and
+that are not referenced by such declarations -/
+meta def all_unneeded : tactic (name_map declaration) :=
+do ds ← local_decls,
+   ls ← ds.keys.mfilter (succeeds ∘ user_attribute.get_param_untyped needed_attr),
+   ds ← ls.mfoldl (flip unneeded) ds,
+   ds.mfilter $ λ n d, do
+     e ← get_env,
+     return $ !d.is_auto_generated e &&
```

Comment:
I can use `is_auto_or_internal` instead of `is_auto_generated` but they don't cover the equations generated by the equation compiler. I can add them to `auto_generated` if that's suitable to its purpose.

Reply:
I'm quite sure that `is_internal` cover all declarations whose name has a part that starts with `_`. So that covers most of your tests. It doesn't directly cover the test `!(n.get_prefix.last = "equations")`, but I believe the last part of those names start with `_`? So `is_auto_or_internal` should catch each of these.

Reply:
@cipher1024 at a glance I think @fpvandoorn is right, do you have an example where this isn't the case?

Reply:
I think he's right too now that I looked at it more closely